### Titanic - Machine Learning from Disaster
My task: create a model that predicts which passengers survived the Titanic shipwreck, using a given dataset

#### Load and Explore the given data set 

In [8]:
!pip install category_encoders

In [5]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from category_encoders import OneHotEncoder
import matplotlib.gridspec as gridspec

In [7]:
data = pd.read_csv('train.csv')

In [9]:
data.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [11]:
data.shape

(891, 12)

My dependent variable is ***survived*** (0 = No, 1 = Yes)

pclass = Ticket class (1 = 1st, 2 = 2nd, 3 = 3rd) \
sibsp - # of siblings / spouses aboard the Titanic \
parch - # of parents / children aboard the Titanic

Categorical variables (need to be transformed): 'Sex' and 'Embarked'

For my first model I will drop the following features (in my opinion they do not have much impact on survival): 'Name', 'Ticket', 'Cabin'

In [33]:
# Delete columns 'Name', 'Ticket', and 'Cabin'

data.drop(['Name', 'Ticket', 'Cabin'], axis=1, inplace = True)

In [35]:
data.head()

PassengerId  Survived  Pclass     Sex   Age  SibSp  Parch     Fare Embarked
0            1         0       3    male  22.0      1      0   7.2500        S
1            2         1       1  female  38.0      1      0  71.2833        C
2            3         1       3  female  26.0      0      0   7.9250        S
3            4         1       1  female  35.0      1      0  53.1000        S
4            5         0       3    male  35.0      0      0   8.0500        S

In [41]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Sex          891 non-null    object 
 4   Age          714 non-null    float64
 5   SibSp        891 non-null    int64  
 6   Parch        891 non-null    int64  
 7   Fare         891 non-null    float64
 8   Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(2)
memory usage: 62.8+ KB


I have 891 - 714 = 177 rows with missed values for the 'Age' variable. I need to decide how to handle this issue. I think 'Age' is a very important variable in this case.


In [49]:
# Convert the categorical variables into binary values

In [67]:
data = pd.get_dummies(data, columns=['Sex', 'Embarked'], dtype=int)

In [69]:
data.head()

PassengerId  Survived  Pclass   Age  SibSp  Parch     Fare  Sex_female  \
0            1         0       3  22.0      1      0   7.2500           0   
1            2         1       1  38.0      1      0  71.2833           1   
2            3         1       3  26.0      0      0   7.9250           1   
3            4         1       1  35.0      1      0  53.1000           1   
4            5         0       3  35.0      0      0   8.0500           0   

   Sex_male  Embarked_C  Embarked_Q  Embarked_S  
0         1           0           0           1  
1         0           1           0           0  
2         0           0           0           1  
3         0           0           0           1  
4         1           0           0           1

In [72]:
data.describe().round(2).T

count    mean     std   min     25%     50%    75%     max
PassengerId  891.0  446.00  257.35  1.00  223.50  446.00  668.5  891.00
Survived     891.0    0.38    0.49  0.00    0.00    0.00    1.0    1.00
Pclass       891.0    2.31    0.84  1.00    2.00    3.00    3.0    3.00
Age          714.0   29.70   14.53  0.42   20.12   28.00   38.0   80.00
SibSp        891.0    0.52    1.10  0.00    0.00    0.00    1.0    8.00
Parch        891.0    0.38    0.81  0.00    0.00    0.00    0.0    6.00
Fare         891.0   32.20   49.69  0.00    7.91   14.45   31.0  512.33
Sex_female   891.0    0.35    0.48  0.00    0.00    0.00    1.0    1.00
Sex_male     891.0    0.65    0.48  0.00    0.00    1.00    1.0    1.00
Embarked_C   891.0    0.19    0.39  0.00    0.00    0.00    0.0    1.00
Embarked_Q   891.0    0.09    0.28  0.00    0.00    0.00    0.0    1.00
Embarked_S   891.0    0.72    0.45  0.00    0.00    1.00    1.0    1.00